src/commonMain/kotlin/org/treeWare/model/operator/Difference.kt

```diff
@@ -0,0 +1,407 @@
+package org.treeWare.model.operator
+
+import org.treeWare.model.core.*
+import org.treeWare.model.traversal.AbstractLeaderManyModelVisitor
+import org.treeWare.model.traversal.TraversalAction
+import org.treeWare.model.traversal.forEach
+import org.treeWare.util.assertInDevMode
+
+
+/*First input should be original, second input should be new*/
+fun difference(inputs: List<MainModel>): List<MutableMainModel?> {
+    assert(inputs.size == 2)
+    val differenceVisitor = DifferenceVisitor()
+    forEach(inputs, differenceVisitor, false)
+
+    /*If a result tree is empty, then instead make the value of the tree null*/
+    val addTree = when{
+        differenceVisitor.stepsFromAdd > 0 -> null
+        else -> differenceVisitor.addMainModel
+    }
+    val deleteTree = when{
+        differenceVisitor.stepsFromDelete > 0 -> null
+        else -> differenceVisitor.deleteMainModel
+    }
+    return listOf(addTree, deleteTree)
+}
+
+private class DifferenceVisitor : AbstractLeaderManyModelVisitor<TraversalAction>(
+    TraversalAction.CONTINUE
+) {
+    var stepsFromAdd = 0 //the number of steps away from the path from the root to an added node
+    var stepsFromDelete = 0 //the number of steps away from the path from the root to a deleted node
+    var stepsFromUpdate = 0 //the number of steps away from the path from the root to an updated node
+    val addStack = ArrayDeque<MutableElementModel>()
```

Comment:
As discussed IRL, use a private `StackState` data class as the elements of your stack. Continue to use 3 separate stacks. The `StackState` class will hold the output model's element, as well as a boolean `var` flag that indicates if the element has children that need to be in the output. This also indicates that the elements needs to be in the output. So the flag could be called something like `includeInOutput`